data/src/main/java/com/hattrick/data/repository/SampleRepositoryImpl.kt

```diff
@@ -1,16 +1,36 @@
 package com.hattrick.data.repository
 
+import android.util.Log
+import com.hattrick.data.model.SampleDataModel
+import com.hattrick.data.model.toDomain
 import com.hattrick.domain.model.SampleModel
 import com.hattrick.domain.repository.SampleRepository
 import kotlinx.coroutines.CoroutineDispatcher
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.withContext
 
-class SampleRepositoryImpl(private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO) :
-    SampleRepository {
+class SampleRepositoryImpl(
+    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
+    private val sampleLocalDataSource: SampleLocalDataSource,
+    private val sampleRemoteDataSource: SampleRemoteDataSource,
+) : SampleRepository {
 
-    override suspend fun sampleFun(): SampleModel = withContext(ioDispatcher) {
-        SampleModel("name")
+    private suspend fun getApiResponse(int: Int): SampleModel? {
+        return try {
+            sampleRemoteDataSource.getApiResponse(int).toDomain()
+        } catch (e: Exception) {
+            Log.e("Error: ", e.toString())
+            null
+        }
+    }
+
+    override suspend fun sampleFun(int: Int): SampleModel = withContext(ioDispatcher) {
+        if (sampleLocalDataSource.getElementFromDb(int) != null) {
+            sampleLocalDataSource.getElementFromDb(int).toDomain()
+        } else {
+            sampleLocalDataSource.saveElementToDb(sampleRemoteDataSource.getApiResponse(int))
+            getApiResponse(int)!!
+        }
```

Comment:
Let’s make this API call once, also avoid using null assertion(!!), in this case let’s throw an exception when the result is null, and use try catch in the repository method call in the domain layer.